python/lsst/cp/pipe/ptc.py

```diff
@@ -0,0 +1,535 @@
+
+    def funcAstier(self, pars, x):
+        """Single brighter-fatter parameter model for PTC; Equation 16 of Astier+19"""
+        a00, gain, noise = pars
+        return 0.5/(a00*gain*gain)*(1 - np.exp(-2*a00*x*gain)) + noise/(gain*gain)
+
+    def fitPtcAndNl(self, fitVectorsDict, ptcFitType='POLYNOMIAL'):
+        """Function to fit PTC, and calculate linearity and linearity residual
+
+        Parameters
+        ----------
+        fitVectorsDicti : `dict`
+            Dictionary with exposure time, mean, and variance vectors in a tuple
+        ptcFitType : `str`
+            Fit a 'POLYNOMIAL' (degree: 'polynomialFitDegree') or '
+            ASTIERAPPROXIMATION' to the PTC
+
+        Returns
+        -------
+        fitPtcDict : `dict`
+            Dictionary of the form fitPtcDict[amp] =
+            (meanVec, varVec, parsFit, parsFitErr, index)
+        nlDict : `dict`
+            Dictionary of the form nlDict[amp] =
+            (timeVec, meanVec, linResidual, parsFit, parsFitErr)
+        """
+        if ptcFitType == 'ASTIERAPPROXIMATION':
+            ptcFunc = self.funcAstier
+            parsIniPtc = [1e-9, 0.7, 10.]
+        if ptcFitType == 'POLYNOMIAL':
+            ptcFunc = self.funcPolynomial
+            parsIniPtc = np.repeat(1., self.config.polynomialFitDegree + 1)
+
+        parsIniNl = [1., 1., 1.]
+        fitPtcDict = {key: {} for key in fitVectorsDict}
+        nlDict = {key: {} for key in fitVectorsDict}
+
+        def errFunc(p, x, y):
+            return ptcFunc(p, x) - y
+
+        maxIterationsPtcOutliers = self.config.maxIterationsPtcOutliers
+        for amp in fitVectorsDict:
+            timeVec, meanVec, varVec = fitVectorsDict[amp]
+            timeVecOriginal = np.array(timeVec)
+            meanVecOriginal = np.array(meanVec)
+            varVecOriginal = np.array(varVec)
+            index = ((meanVecOriginal > self.config.minMeanSignal) &
+                     (meanVecOriginal <= self.config.maxMeanSignal))
+            initialPoints = np.count_nonzero(index)
+            #  Before bootstrap fit, do an iterative fit to get rid of outliers in PTC
+            count = 1
+            sigmaCutPtcOutliers = self.config.sigmaCutPtcOutliers
+            maxIterationsPtcOutliers = self.config.maxIterationsPtcOutliers
+            while count <= maxIterationsPtcOutliers:
+                pars, cov = leastsq(errFunc, parsIniPtc, args=(meanVecOriginal[index],
+                                    varVecOriginal[index]), full_output=0)
+                sigResids = (varVecOriginal[index] - ptcFunc(pars,
+                             meanVecOriginal[index]))/np.sqrt(varVecOriginal[index])
+                index = list(np.where(np.abs(sigResids) < sigmaCutPtcOutliers)[0])
+                self.log.info((f"PTC outliers in amplifier {amp}: {initialPoints - len(index)} out of"
+                               f"{initialPoints} at iteration {count} out of {maxIterationsPtcOutliers}"))
+                count += 1
+
```

Comment:
The plot I'm looking at (ptc_davis_data_e2v_aug23_cubic.pdf) seems to be including bad points for C0[4-7], resulting in large errors on the noise and gain parameters.  I just wanted to check that this is expected/wasn't run with the iterative rejection/etc.

Reply:
Yes, that's correct